feat: Expose Rev Analytics views for Max SQL queries

rafaeelaudibert opened the pull request:
Max only knows about a handful of built-in SQL tables + the tables coming from the data warehouse. Let's make it aware of our revenue analytics views. We can always revert this if we see worse performance from Max.

Closes #32486

kappa90 commented on ee/hogai/graph/sql/nodes.py:
```diff
@@ -52,11 +53,9 @@ def run(self, state: AssistantState, config: RunnableConfig) -> PartialAssistant
                 for table_name, table in serialized_database.items()
-                # Only the most important core tables, plus all warehouse tables
-                if table_name in ["events", "groups", "persons"] or table_name in database.get_warehouse_tables()
+                if self._should_include_table_in_schema(table_name)
```
+1 to what Greptile said.

rafaeelaudibert replied:
Ah, I refactored this at the end and missed this bit

rafaeelaudibert replied:
BTW, any reason why we don't expose all tables/views to Max? Are we trying to avoid spending way too much context/tokens/money on useless tables?

What if I tweaked the code to let customers choose which views Max should know about and which ones it shouldn't?

kappa90 replied:
I think we want all tables to be available in Max. If we're not doing that, it's a mistake.

kappa90 commented on posthog/schema.py:
```diff
@@ -8826,6 +8826,15 @@ class DatabaseSchemaManagedViewTable(BaseModel):
 
+    def description(self) -> Optional[str]:
+        if self.kind == DatabaseSchemaManagedViewTableKind.REVENUE_ANALYTICS_CHARGE:
+            return "Useful for Revenue-related questions, contains charges for customers"
```
"charges for customers" -> "charges per customer"?

rafaeelaudibert replied:
Yeah, not really. It's all the charges we've ever received, and they *might* map to a customer. Will work on improving this

kappa90 reviewed (changes requested):
Did you test this and Max was actually able to respond to revenue based questions?

There's a bug that Greptile caught that I think breaks the functionality.

The way Max works right now is that it decides what kind of insight it wants to generate (trends, funnel, retention or SQL) and only AFTER explores what kind of data it can retrieve.

So if revenue data is always based on DWH tables, we need to explicitly tell Max about it.

There's a prompt in `ee/hogai/graph/root/prompts.py` called `ROOT_SYSTEM_PROMPT` which tells Max which insight to go for.

Specifically, for SQL, it says:

```
## 'sql'

The 'sql' insight type allows you to write arbitrary SQL queries to retrieve data.

The SQL insights have the following features:
- Filter data using arbitrary SQL.
- All ClickHouse SQL features.
- You can nest subqueries as needed.
```

I think it would be worth it to mention there that all revenue-based data exists in tables that can only be queried via SQL.

skoob13 commented on posthog/schema.py:
```diff
@@ -8826,6 +8826,15 @@ class DatabaseSchemaManagedViewTable(BaseModel):
     source_id: Optional[str] = None
     type: Literal["managed_view"] = "managed_view"
 
+    def description(self) -> Optional[str]:
```
This file is autogenerated by the `pnpm run schema:build` command so we don't manually extend it. You can create a mapping somewhere else, like in the node code.

rafaeelaudibert replied:
Ah, my bad, yes

rafaeelaudibert commented:
@kappa90 

> I think it would be worth it to mention there that all revenue-based data exists in tables that can only be queried via SQL.

That's not correct, though. You can use them via insights, funnels, whatever. It's just another view. I'll make sure I add some more information to the base prompt to point Max towards the revenue views.

---

Whether I've tested it locally: this was just an early draft that I released on Friday, I couldn't get it to work locally, working on it right now, will report back. Thank you :)

rafaeelaudibert commented:
Even tho Im actively asking it to use SQL, it doesn't 🤷🏻 

![image](https://github.com/user-attachments/assets/e06fe3e0-7766-4eaa-b58f-6c319312c95e)

rafaeelaudibert committed:
feat: Expose Rev Analytics views for Max SQL queries

Max only knows about a handful of built-in SQL tables + the tables coming from the data warehouse. Let's make it aware of our revenue analytics views. We can always revert this if we see worse performance from Max.

kappa90 commented:
> Even tho Im actively asking it to use SQL, it doesn't 🤷🏻
> 
> ![image](https://private-user-images.githubusercontent.com/32079912/447654945-e06fe3e0-7766-4eaa-b58f-6c319312c95e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDgyODc1OTcsIm5iZiI6MTc0ODI4NzI5NywicGF0aCI6Ii8zMjA3OTkxMi80NDc2NTQ5NDUtZTA2ZmUzZTAtNzc2Ni00ZWFhLWI1OGYtNmMzMTkzMTJjOTVlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA1MjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNTI2VDE5MjEzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTcyNDc0N2NmYzU3NWVhYzQ1YzE4OGQ4NTk0ZDBjNTk2ZjE0M2Y2Nzc4ZTEzNjljZGU0YmNhMmZjYmYyOGYwNDgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.HUdSsT4UjOzsdvLVn8Ce3psoQJI0GGuRScLwJikNj64)

I'll test this and come back to you!

skoob13 commented on ee/hogai/graph/sql/nodes.py:
```diff
@@ -75,6 +74,28 @@ def run(self, state: AssistantState, config: RunnableConfig) -> PartialAssistant
         )
         return super()._run_with_prompt(state, prompt, config=config)
 
+    def _schema_description_for_table(self, table_name: str, table: DatabaseSchemaTable) -> str:
```
```suggestion
    def _get_schema_description_for_table(self, table_name: str, table: DatabaseSchemaTable) -> str:
```

skoob13 commented on ee/hogai/graph/sql/nodes.py:
```diff
@@ -75,6 +74,28 @@ def run(self, state: AssistantState, config: RunnableConfig) -> PartialAssistant
+
+        return schema_description
+
+    def _should_include_table_in_schema(self, table_name: str, table: DatabaseSchemaTable) -> bool:
```
This condition always returns false for revenue analytics views. I tried both a Stripe source and an event source, but neither worked. Did we miss something here?

skoob13 commented on ee/hogai/graph/sql/prompts.py:
```diff
@@ -38,3 +42,11 @@
+def description_for_table(table: DatabaseSchemaTable) -> Optional[str]:
+    if isinstance(table, DatabaseSchemaManagedViewTable):
+        if table.kind == DatabaseSchemaManagedViewTableKind.REVENUE_ANALYTICS_CHARGE:
+            return "Useful for Revenue-related questions, contains charges for customers"
```
I'd describe the view more generically so we don't limit the agent to revenue-related questions. For example, calculating CAC is not directly related to revenue, but the metric needs a revenue value to be calculated. It would also be nice to describe the data relationship between the two revenue analytics views. I assume they're related, correct?

skoob13 commented on ee/hogai/graph/sql/test/test_nodes.py:
```diff
@@ -8,11 +8,26 @@
 from posthog.test.base import BaseTest
 
 
+def deindent(text: str) -> str:
```
Doesn't`textwrap.dedent` do the same job?

skoob13 reviewed (changes requested):
Left a few suggestions of improvement, but I couldn't make it work. Do I miss some DWH setup?
